### DISCUSSION TITLE
How to use a generic interface as store type?

### DISCUSSION THREAD

**Discussion by xypnox** [pmndrs/zustand#841] [Q&A] [answered]
Hi,

I have a generic interface:

```ts
interface StoreState<Item> {
  items: Item[]
  getItems: () => Item[] 
//...
}
``` 

How can I create a zustand store using it as the store type?

I have tried `const useSearchViewStore = create<SearchViewState<Item>>((set) => (/*...*/))` but typescript error says:
`Cannot find name Item`

**Comment by dai-shi** [comment, marked as answer]
How about this?
https://codesandbox.io/s/tender-bell-95sko3?file=/src/App.tsx

```ts
interface StoreState<Item> {
  items: Item[];
}

const useStoreBase = create<StoreState<any>>((set) => ({
  items: []
}));

export const useStore = <Item, Slice>(
  selector: (state: StoreState<Item>) => Slice
) => useStoreBase(selector);
```

**Reply by xypnox** [reply]
While selecting a state inside via the useStore, the type of the items fetched gets unset to `unknown[]`.

Currently I am adding `as Item[]` after the hook call for correct type.

Is there a way to preserve the type?

**Reply by dai-shi** [reply]
`unknown[]` sounds correct, because you don't know `Item[]` at the point of creation, no?

So, maybe the correct answer is, store can't have generics, because the store is not a function but a value.

**Reply by xypnox** [reply]
Makes sense.

I have replaced the `any` to `unknown` in my code to better reflect it.

Thanks for clarifying.

**Reply by AndresdoSantos** [reply]
Hi, I know ```Slice``` is a generic one, but how should it be used? Is it the return from the ```useStore``` hook?

I did it as a return from the ```useStore``` hook and I got the error:

**An argument for 'selector' was not provided.**

**Reply by dai-shi** [reply]
Can you provide a code snippet or a small codesandbox repro?

**Reply by AndresdoSantos** [reply]
Thanks for helping me.

Before showing the code I will say what I expect from it:

- I want a hook that provides data to a table

The dynamic is like this:

1. Bring in the API data
2. Repost them to the table
3. Display them in the table

Which files use generic types?

```Table.tsx```
```useData.ts```
```get.ts```

What is the problem I have?

- I can't pass types to ```useData``` preventing me from having the data correctly inside the main component.

Here is the codeSandbox link: https://codesandbox.io/s/generic-types-in-zustand-y79mij

**Reply by dai-shi** [reply]
There's no magic. You need to assert types: https://codesandbox.io/s/generic-types-in-zustand-forked-qd5lg7?file=/src/hooks/useData.ts
Note that we are making types to lie. (runtime errors can happen.)

So, I would prefer type assertion on use: https://codesandbox.io/s/generic-types-in-zustand-forked-2qyr0c?file=/src/App.tsx

**Reply by AndresdoSantos** [reply]
Thanks!

**Comment by moklick** [comment]
Hey @dai-shi ,

thanks this is already helpful but I can't get it to work. In this codesandbox I am using your example to specify the types of the store items. As soon as I pass a generic for items, the returned type is not correct anymore. Adjusted example:

https://codesandbox.io/p/sandbox/modest-blackwell-m3mhxx?file=%2Fsrc%2FApp.tsx%3A20%2C1-21%2C1

It seems that I have a misunderstand of TS in this case but I can't figure out what's wrong.

**Reply by dai-shi** [reply]
Does this help? https://tsplay.dev/NnoQvm

```tsx
import { create } from "zustand";
import "./styles.css";

interface StoreState<Item> {
  items: Item[];
}

const useStoreBase = create<StoreState<any>>((set) => ({
  items: [],
}));

type StoreItem = { value: number };

export const useStore = <Item, Slice>(
  selector: (state: StoreState<Item>) => Slice,
) => useStoreBase(selector);

export default function App() {
  const items = useStore((s: StoreState<StoreItem>) => s.items);
  //    ^?

  return (
    <div className="App">
      <h1>Hello CodeSandbox</h1>
      <h2>Start editing to see some magic happen!</h2>
      <div>{JSON.stringify(items)}</div>
    </div>
  );
}
```

**Reply by dai-shi** [reply]
Or this: https://tsplay.dev/NnoQvm

```ts
import { create, StoreApi, UseBoundStore } from "zustand";
import "./styles.css";

interface StoreState<Item> {
  items: Item[];
}

const useStoreBase = create<StoreState<any>>((set) => ({
  items: [],
}));

type StoreItem = { value: number };

export const useStore: UseBoundStore<StoreApi<StoreState<StoreItem>>> =
  useStoreBase;

export default function App() {
  const items = useStore((s: StoreState<StoreItem>) => s.items);
  //    ^?

  return (
    <div className="App">
      <h1>Hello CodeSandbox</h1>
      <h2>Start editing to see some magic happen!</h2>
      <div>{JSON.stringify(items)}</div>
    </div>
  );
}
```

**Reply by moklick** [reply]
thanks a lot! I got it now :) The main issue was, that I didn't know that TS infers all generics or none. I wanted the first generic (StoreItem) to be passed by the user and the second (StateSlice) inferred. As explained in your solution, it's better to let the user pass the Item type to the selector.
